agr_java_core/src/main/java/org/alliancegenome/core/translators/tdf/AlleleToTdfTranslator.java

```diff
@@ -30,24 +30,50 @@ public String getAllRows(List<Allele> annotations) {
 
     public List<AlleleDownloadRow> getAlleleDownloadRowsForGenes(List<Allele> annotations) {
 
+denormalizeAnnotations(annotations);
         return annotations.stream()
 
+                .filter(annotation -> annotation.getVariants()!=null)
+                .filter(annotation -> !CollectionUtils.isEmpty(annotation.getVariants()))
                 .map(annotation -> annotation.getVariants().stream()
-                        .map(entity -> entity.getPublications().stream()
-                                .map(pub -> {
-                                    return List.of(getBaseDownloadRow(annotation, entity, pub));
-                                })
-                                .flatMap(Collection::stream)
-                                .collect(Collectors.toList()))
-                        .flatMap(Collection::stream)
+                        .map(join -> {
+                            AlleleDownloadRow row = getBaseDownloadRow(annotation, join, null);
+                            return row;
+                        })
                         .collect(Collectors.toList()))
                 .flatMap(Collection::stream)
                 .collect(Collectors.toList());
-
     }
 
+
+        private void denormalizeAnnotations(List<Allele> alleles) {
+            // add genetic entity info for annotations with pure genes
+
+
+
+            alleles.stream()
+                    .filter(annotation -> CollectionUtils.isEmpty(annotation.getVariants()))
+                    .forEach(annotation -> {
+                        Variant entity = createNewVariant();
```

Comment:
Why do you have to create an empty Variant object? Can you handle the null at the other end where you are expecting a non-null Variant object?